File: week04_mcts/alphano/mcts_agent.cpp
#include <bits/stdc++.h>
using namespace std;

int board[8][8];
int turn;

const int dx[8] = {-1,-1,-1,0,0,1,1,1};
const int dy[8] = {-1,0,1,-1,1,-1,0,1};

int root_player;

void copy_board(int src[8][8], int dst[8][8]) {
    for (int i = 1; i <= 7; i++) {
        for (int j = 1; j <= 7; j++) {
            dst[i][j] = src[i][j];
        }
    }
}

int count_pieces(int b[8][8], int player) {
    int cnt = 0;
    for (int i = 1; i <= 7; i++) {
        for (int j = 1; j <= 7; j++) {
            if (b[i][j] == player) cnt++;
        }
    }
    return cnt;
}

void sim_move(int b[8][8], int x1, int y1, int x2, int y2, int player) {
    int dist = max(abs(x2 - x1), abs(y2 - y1));

    if (dist == 1) {
        // Split move
        b[x2][y2] = player;
    } else {
        // Jump move
        b[x1][y1] = 0;
        b[x2][y2] = player;
    }

    // Infect adjacent opponent pieces
    for (int d = 0; d < 8; d++) {
        int nx = x2 + dx[d];
        int ny = y2 + dy[d];
        if (nx >= 1 && nx <= 7 && ny >= 1 && ny <= 7) {
            if (b[nx][ny] == (player ^ 3)) {
                b[nx][ny] = player;
            }
        }
    }
}

int gen_moves(int b[8][8], int player, int moves[][4]) {
    int cnt = 0;
    for (int x1 = 1; x1 <= 7; x1++) {
        for (int y1 = 1; y1 <= 7; y1++) {
            if (b[x1][y1] != player) continue;

            for (int x2 = x1 - 2; x2 <= x1 + 2; x2++) {
                if (x2 < 1 || x2 > 7) continue;

                for (int y2 = y1 - 2; y2 <= y1 + 2; y2++) {
                    if (y2 < 1 || y2 > 7) continue;
                    if (x2 == x1 && y2 == y1) continue;
                    if (b[x2][y2] != 0) continue;

                    moves[cnt][0] = x1;
                    moves[cnt][1] = y1;
                    moves[cnt][2] = x2;
                    moves[cnt][3] = y2;
                    cnt++;
                }
            }
        }
    }
    return cnt;
}

struct Node {
    int b[8][8];
    int player;
    Node* parent;
    int move[4];
    vector<Node*> children;
    double wins;
    int visits;
    int untried[200][4];
    int untried_cnt;

    Node(int board[8][8], int p, Node* par, int* m) {
        copy_board(board, b);
        player = p;
        parent = par;
        if (m) {
            move[0] = m[0];
            move[1] = m[1];
            move[2] = m[2];
            move[3] = m[3];
        }
        wins = 0.0;
        visits = 0;
        untried_cnt = gen_moves(b, player, untried);
    }

    ~Node() {
        for (Node* child : children) {
            delete child;
        }
    }

    bool is_terminal() {
        return untried_cnt == 0 && children.empty();
    }

    bool has_untried_moves() {
        return untried_cnt > 0;
    }
};

double ucb1(Node* n) {
    if (n->visits == 0) return 1e9;
    return (n->wins / n->visits) + 1.414 * sqrt(log(n->parent->visits) / n->visits);
}

Node* best_child_ucb(Node* node) {
    Node* best = nullptr;
    double best_score = -1e9;

    for (Node* child : node->children) {
        double score = ucb1(child);
        if (score > best_score) {
            best_score = score;
            best = child;
        }
    }

    return best;
}

Node* expand(Node* node) {
    int idx = rand() % node->untried_cnt;
    int* m = node->untried[idx];

    int new_board[8][8];
    copy_board(node->b, new_board);
    sim_move(new_board, m[0], m[1], m[2], m[3], node->player);

    Node* child = new Node(new_board, node->player ^ 3, node, m);
    node->children.push_back(child);

    // Remove from untried
    node->untried[idx][0] = node->untried[node->untried_cnt - 1][0];
    node->untried[idx][1] = node->untried[node->untried_cnt - 1][1];
    node->untried[idx][2] = node->untried[node->untried_cnt - 1][2];
    node->untried[idx][3] = node->untried[node->untried_cnt - 1][3];
    node->untried_cnt--;

    return child;
}

Node* tree_policy(Node* root) {
    Node* node = root;

    while (!node->is_terminal()) {
        if (node->has_untried_moves()) {
            return expand(node);
        } else {
            node = best_child_ucb(node);
        }
    }

    return node;
}

double rollout(Node* node) {
    int sim_board[8][8];
    copy_board(node->b, sim_board);
    int current_player = node->player;

    for (int step = 0; step < 200; step++) {
        int moves[200][4];
        int cnt = gen_moves(sim_board, current_player, moves);

        if (cnt == 0) {
            // Pass
            current_player ^= 3;
            int opp_cnt = gen_moves(sim_board, current_player, moves);
            if (opp_cnt == 0) {
                // Both pass - game over
                break;
            }
            continue;
        }

        // Random move
        int idx = rand() % cnt;
        sim_move(sim_board, moves[idx][0], moves[idx][1], moves[idx][2], moves[idx][3], current_player);
        current_player ^= 3;
    }

    // Evaluate from root player's perspective
    int my_count = count_pieces(sim_board, root_player);
    int opp_count = count_pieces(sim_board, root_player ^ 3);

    if (my_count > opp_count) return 1.0;
    if (my_count < opp_count) return 0.0;
    return 0.5;
}

void backprop(Node* node, double result) {
    while (node) {
        node->visits++;
        // Result is from root player's perspective
        // node->player is who plays NEXT from this state
        // If it's opponent's turn at this node, this is good for root player
        if (node->player != root_player) {
            node->wins += result;
        } else {
            node->wins += (1.0 - result);
        }
        node = node->parent;
    }
}

tuple<int,int,int,int> mcts_search(int time_limit_ms) {
    Node* root = new Node(board, turn, nullptr, nullptr);
    root_player = turn;

    auto start = chrono::high_resolution_clock::now();
    int iterations = 0;

    while (true) {
        auto now = chrono::high_resolution_clock::now();
        auto elapsed = chrono::duration_cast<chrono::milliseconds>(now - start).count();
        if (elapsed >= time_limit_ms) break;

        Node* node = tree_policy(root);
        double result = rollout(node);
        backprop(node, result);
        iterations++;
    }

    // Best child = most visited
    Node* best = nullptr;
    int max_visits = -1;
    for (Node* child : root->children) {
        if (child->visits > max_visits) {
            max_visits = child->visits;
            best = child;
        }
    }

    cerr << "MCTS: " << iterations << " iterations, best visits: " << max_visits << endl;

    tuple<int,int,int,int> result;
    if (best) {
        result = make_tuple(best->move[0], best->move[1], best->move[2], best->move[3]);
    } else {
        // No moves found - should not happen if called correctly
        result = make_tuple(-1, -1, -1, -1);
    }

    delete root;
    return result;
}

tuple<int,int,int,int> find_move(int my_time) {
    int time_limit;
    if (my_time > 30000) time_limit = 800;
    else if (my_time > 10000) time_limit = 400;
    else if (my_time > 3000) time_limit = 150;
    else time_limit = 50;

    return mcts_search(time_limit);
}

void apply_my_move(int x1, int y1, int x2, int y2) {
    if (x1 == -1) return; // pass

    int dist = max(abs(x2 - x1), abs(y2 - y1));
    assert(dist == 1 || dist == 2);
    assert(board[x1][y1] == turn);
    assert(board[x2][y2] == 0);

    if (dist == 1) {
        board[x2][y2] = turn;
    } else {
        board[x1][y1] = 0;
        board[x2][y2] = turn;
    }

    for (int d = 0; d < 8; d++) {
        int nx = x2 + dx[d];
        int ny = y2 + dy[d];
        if (nx >= 1 && nx <= 7 && ny >= 1 && ny <= 7) {
            if (board[nx][ny] == (turn ^ 3)) {
                board[nx][ny] = turn;
            }
        }
    }
}

void apply_opp_move(int x1, int y1, int x2, int y2) {
    if (x1 == -1) return; // pass

    int opp = turn ^ 3;
    int dist = max(abs(x2 - x1), abs(y2 - y1));
    assert(dist == 1 || dist == 2);
    assert(board[x1][y1] == opp);
    assert(board[x2][y2] == 0);

    if (dist == 1) {
        board[x2][y2] = opp;
    } else {
        board[x1][y1] = 0;
        board[x2][y2] = opp;
    }

    for (int d = 0; d < 8; d++) {
        int nx = x2 + dx[d];
        int ny = y2 + dy[d];
        if (nx >= 1 && nx <= 7 && ny >= 1 && ny <= 7) {
            if (board[nx][ny] == turn) {
                board[nx][ny] = opp;
            }
        }
    }
}

int main() {
    srand(time(0));

    board[1][1] = board[7][7] = 1;
    board[1][7] = board[7][1] = 2;

    string line;
    while (getline(cin, line)) {
        istringstream in(line);
        string cmd;
        in >> cmd;

        if (cmd == "READY") {
            string role;
            in >> role;
            turn = (role == "FIRST") ? 1 : 2;
            cout << "OK" << endl;
        } else if (cmd == "TURN") {
            int t1, t2;
            in >> t1 >> t2;

            auto [x1, y1, x2, y2] = find_move(t1);
            apply_my_move(x1, y1, x2, y2);
            cout << "MOVE " << x1 << " " << y1 << " " << x2 << " " << y2 << endl;
        } else if (cmd == "OPP") {
            int x1, y1, x2, y2, t;
            in >> x1 >> y1 >> x2 >> y2 >> t;
            apply_opp_move(x1, y1, x2, y2);
        } else if (cmd == "FINISH") {
            break;
        }
    }

    return 0;
}

File: week04_mcts/alphano/mcts_agent.py
#!/usr/bin/env python3
"""
ALPHANO ATAXX MCTS Agent
Week 4: Monte Carlo Tree Search

ALPHANO 프로토콜:
- READY FIRST/SECOND → print("OK")
- TURN my_time opp_time → print("MOVE x1 y1 x2 y2") (1-indexed)
- OPP x1 y1 x2 y2 → 보드 업데이트
- FINISH → 종료
"""

import sys
import time
import random
import math
from copy import deepcopy


class AtaxxBoard:
    """ATAXX 7x7 보드"""

    def __init__(self):
        """보드 초기화"""
        self.board = [[0] * 7 for _ in range(7)]
        # 초기 배치 (0-indexed)
        self.board[0][0] = 1  # FIRST
        self.board[6][6] = 1
        self.board[0][6] = 2  # SECOND
        self.board[6][0] = 2
        self.current_player = 1  # 1=FIRST, 2=SECOND

    def copy(self):
        """보드 복사"""
        new_board = AtaxxBoard()
        new_board.board = deepcopy(self.board)
        new_board.current_player = self.current_player
        return new_board

    def get_legal_moves(self):
        """가능한 모든 수 반환 (1-indexed)"""
        moves = []
        player = self.current_player

        for r in range(7):
            for c in range(7):
                if self.board[r][c] == player:
                    # Split (거리 1): 8방향
                    for dr in [-1, 0, 1]:
                        for dc in [-1, 0, 1]:
                            if dr == 0 and dc == 0:
                                continue
                            nr, nc = r + dr, c + dc
                            if 0 <= nr < 7 and 0 <= nc < 7 and self.board[nr][nc] == 0:
                                # (r, c)에서 (nr, nc)로 Split
                                moves.append((r + 1, c + 1, nr + 1, nc + 1))

                    # Jump (거리 2): 8방향
                    for dr in [-2, -1, 0, 1, 2]:
                        for dc in [-2, -1, 0, 1, 2]:
                            if abs(dr) <= 1 and abs(dc) <= 1:
                                continue  # 거리 1은 이미 처리
                            if max(abs(dr), abs(dc)) > 2:
                                continue  # 거리 2 초과
                            nr, nc = r + dr, c + dc
                            if 0 <= nr < 7 and 0 <= nc < 7 and self.board[nr][nc] == 0:
                                # (r, c)에서 (nr, nc)로 Jump
                                moves.append((r + 1, c + 1, nr + 1, nc + 1))

        # 중복 제거
        moves = list(set(moves))

        # 가능한 수가 없으면 PASS
        if not moves:
            moves = [None]

        return moves

    def apply_move(self, move):
        """수를 적용 (원본 변경하지 않고 새 보드 반환)"""
        new_board = self.copy()
        new_board._apply_move_inplace(move)
        return new_board

    def _apply_move_inplace(self, move):
        """수를 적용 (원본 변경)"""
        if move is None:
            # PASS
            self.current_player = 3 - self.current_player
            return

        r1, c1, r2, c2 = move
        r1 -= 1
        c1 -= 1
        r2 -= 1
        c2 -= 1  # 0-indexed로 변환

        player = self.current_player

        # 거리 계산
        dist = max(abs(r2 - r1), abs(c2 - c1))

        if dist == 1:
            # Split: 복제
            self.board[r2][c2] = player
        elif dist == 2:
            # Jump: 이동
            self.board[r1][c1] = 0
            self.board[r2][c2] = player
        else:
            raise ValueError(f"Invalid move distance: {dist}")

        # 인접 8방향 감염
        for dr in [-1, 0, 1]:
            for dc in [-1, 0, 1]:
                if dr == 0 and dc == 0:
                    continue
                nr, nc = r2 + dr, c2 + dc
                if 0 <= nr < 7 and 0 <= nc < 7:
                    if self.board[nr][nc] == (3 - player):
                        self.board[nr][nc] = player

        # 턴 전환
        self.current_player = 3 - self.current_player

    def is_terminal(self):
        """게임 종료 확인"""
        # 돌이 하나도 없으면 종료
        count1 = sum(row.count(1) for row in self.board)
        count2 = sum(row.count(2) for row in self.board)

        if count1 == 0 or count2 == 0:
            return True

        # 양쪽 다 수를 둘 수 없으면 종료
        moves = self.get_legal_moves()
        if moves == [None]:
            # 상대도 확인
            self.current_player = 3 - self.current_player
            opp_moves = self.get_legal_moves()
            self.current_player = 3 - self.current_player

            if opp_moves == [None]:
                return True

        # 보드가 꽉 차면 종료
        empty = sum(row.count(0) for row in self.board)
        if empty == 0:
            return True

        return False

    def get_result(self):
        """
        현재 플레이어 관점의 게임 결과

        Returns:
            1.0: 현재 플레이어 승리
            0.0: 현재 플레이어 패배
            0.5: 무승부
        """
        count1 = sum(row.count(1) for row in self.board)
        count2 = sum(row.count(2) for row in self.board)

        if count1 > count2:
            winner = 1
        elif count2 > count1:
            winner = 2
        else:
            return 0.5  # 무승부

        return 1.0 if winner == self.current_player else 0.0

    def count_pieces(self):
        """돌 개수 세기"""
        count1 = sum(row.count(1) for row in self.board)
        count2 = sum(row.count(2) for row in self.board)
        return count1, count2


class MCTSNode:
    """Monte Carlo Tree Search 노드"""

    def __init__(self, state, parent=None, move=None):
        """
        Args:
            state: AtaxxBoard 상태
            parent: 부모 노드
            move: 부모에서 이 노드로 오는 수
        """
        self.state = state
        self.parent = parent
        self.move = move
        self.children = []
        self.wins = 0.0
        self.visits = 0
        self.untried_moves = state.get_legal_moves()

    def is_fully_expanded(self):
        """모든 자식이 확장되었는가?"""
        return len(self.untried_moves) == 0

    def is_terminal(self):
        """터미널 노드인가?"""
        return self.state.is_terminal()

    def ucb1(self, c=1.414):
        """
        UCB1 값 계산

        Args:
            c: 탐험 상수 (exploration constant)

        Returns:
            UCB1 값
        """
        if self.visits == 0:
            return float('inf')  # 미방문 노드는 최우선

        # Exploitation: 평균 승률
        exploitation = self.wins / self.visits

        # Exploration: 불확실성 보너스
        exploration = c * math.sqrt(math.log(self.parent.visits) / self.visits)

        return exploitation + exploration

    def select_child(self):
        """UCB1 값이 가장 큰 자식 선택"""
        return max(self.children, key=lambda child: child.ucb1())

    def expand(self):
        """
        미탐색 수 중 하나를 선택해서 자식 노드 추가

        Returns:
            새로 추가된 자식 노드
        """
        move = self.untried_moves.pop()
        next_state = self.state.apply_move(move)
        child = MCTSNode(next_state, parent=self, move=move)
        self.children.append(child)
        return child

    def rollout(self):
        """
        현재 상태에서 게임 끝까지 랜덤 플레이

        Returns:
            게임 결과 (현재 플레이어 관점)
        """
        state = self.state.copy()

        # 게임이 끝날 때까지 랜덤 플레이
        while not state.is_terminal():
            legal_moves = state.get_legal_moves()
            move = random.choice(legal_moves)
            state._apply_move_inplace(move)

        # 결과를 원래 플레이어 관점으로 변환
        # state.current_player는 게임 종료 후 플레이어
        # self.state.current_player는 롤아웃 시작 시 플레이어

        count1, count2 = state.count_pieces()

        if count1 > count2:
            winner = 1
        elif count2 > count1:
            winner = 2
        else:
            return 0.5  # 무승부

        # 원래 플레이어 관점으로 결과 반환
        return 1.0 if winner == self.state.current_player else 0.0

    def backpropagate(self, result):
        """
        결과를 루트까지 역전파

        Args:
            result: 시뮬레이션 결과
        """
        node = self
        while node is not None:
            node.visits += 1
            node.wins += result
            result = 1.0 - result  # 관점 전환 (부모는 상대편)
            node = node.parent

    def best_child(self, c=0):
        """
        최선의 자식 선택

        Args:
            c: 0이면 방문 횟수 기준, 0 아니면 UCB1 기준

        Returns:
            최선의 자식 노드
        """
        if c == 0:
            # 방문 횟수가 가장 많은 자식
            return max(self.children, key=lambda child: child.visits)
        else:
            # UCB1 값이 가장 큰 자식
            return max(self.children, key=lambda child: child.ucb1(c))


def mcts_search(board, time_limit_ms):
    """
    Monte Carlo Tree Search로 최선의 수 찾기

    Args:
        board: AtaxxBoard 상태
        time_limit_ms: 시간 제한 (밀리초)

    Returns:
        최선의 수 (1-indexed tuple) 또는 None (PASS)
    """
    root = MCTSNode(board)
    start_time = time.time()
    end_time = start_time + time_limit_ms / 1000.0

    iterations = 0

    while time.time() < end_time:
        node = root

        # 1. Selection: 리프 노드까지 내려가기
        while not node.is_terminal() and node.is_fully_expanded():
            node = node.select_child()

        # 2. Expansion: 가능하면 확장
        if not node.is_terminal() and not node.is_fully_expanded():
            node = node.expand()

        # 3. Simulation: 게임 끝까지 랜덤 플레이
        result = node.rollout()

        # 4. Backpropagation: 결과 역전파
        node.backpropagate(result)

        iterations += 1

    # 디버깅 정보 출력 (stderr로 출력하면 ALPHANO가 무시)
    print(f"MCTS: {iterations} iterations in {time_limit_ms}ms", file=sys.stderr)

    # 최종 수 선택: 가장 많이 방문된 자식
    if root.children:
        best = root.best_child(c=0)
        print(f"Best move: {best.move}, visits: {best.visits}, winrate: {best.wins / best.visits:.2%}", file=sys.stderr)
        return best.move
    else:
        # 가능한 수가 없음 (PASS)
        return None


def main():
    """ALPHANO 프로토콜 메인 루프"""
    board = AtaxxBoard()
    my_color = None

    while True:
        try:
            line = input().strip()
            parts = line.split()

            if not parts:
                continue

            command = parts[0]

            if command == "READY":
                # READY FIRST/SECOND
                color = parts[1]
                if color == "FIRST":
                    my_color = 1
                elif color == "SECOND":
                    my_color = 2
                print("OK")
                sys.stdout.flush()

            elif command == "TURN":
                # TURN my_time opp_time
                my_time = int(parts[1])
                opp_time = int(parts[2])

                # 시간 관리: 남은 시간에 따라 조절
                if my_time > 1000:
                    time_limit = 150  # 1초 이상 남았으면 150ms 사용
                else:
                    time_limit = 10   # 시간이 없으면 10ms만 사용

                # MCTS로 수 찾기
                best_move = mcts_search(board, time_limit)

                if best_move is None:
                    # PASS (가능한 수가 없음)
                    print("PASS")
                else:
                    # MOVE x1 y1 x2 y2
                    print(f"MOVE {best_move[0]} {best_move[1]} {best_move[2]} {best_move[3]}")

                sys.stdout.flush()

                # 보드 업데이트
                if best_move is not None:
                    board._apply_move_inplace(best_move)

            elif command == "OPP":
                # OPP x1 y1 x2 y2 (상대방 수)
                if len(parts) == 5:
                    x1, y1, x2, y2 = map(int, parts[1:5])
                    move = (x1, y1, x2, y2)
                    board._apply_move_inplace(move)
                elif parts[1] == "PASS":
                    # 상대방 PASS
                    board.current_player = 3 - board.current_player

            elif command == "FINISH":
                # 게임 종료
                break

        except EOFError:
            break
        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
            break


if __name__ == "__main__":
    main()
